src/main/java/net/minecraftforge/event/GrindstoneEvent.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.event;
+
+import net.minecraft.world.entity.player.Inventory;
+import net.minecraft.world.inventory.GrindstoneMenu;
+import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.eventbus.api.Cancelable;
+import net.minecraftforge.eventbus.api.Event;
+
+public abstract class GrindstoneEvent extends Event
+{
+    private final ItemStack top;
+    private final ItemStack bottom;
+    private int xp;
+
+    protected GrindstoneEvent(ItemStack top, ItemStack bottom, int xp)
+    {
+        this.top = top;
+        this.bottom = bottom;
+        this.xp = xp;
+    }
+
+    /**
+     * @return The item in the top input grindstone slot. <br>
+     */
+    public ItemStack getTop()
+    {
+        return top;
+    }
+
+    /**
+     * @return The item in the bottom input grindstone slot. <br>
+     */
+    public ItemStack getBottom()
+    {
+        return bottom;
+    }
+
+    /**
+     * This is the experience amount determined by the event, not the vanilla behavior. <br>
+     * If you are the first receiver of this event, it is guaranteed to be -1. <br>
+     * The value must be set to a positive value to override vanilla behavior. <br>
+     * if the value is equal to -1, the vanilla behavior for calculating experience will run. <br>
+     * @return The experience amount given to the player. <br>
+     */
+    public int getXp()
+    {
+        return xp;
+    }
+
+    /**
+     * Sets the experience amount. <br>
+     * @param xp The experience amount given to the player. <br>
+     */
+    public void setXp(int xp)
+    {
+        this.xp = xp;
+    }
+
+    /**
+     * {@link OnplaceItem} is fired when the inputs to a grindstone are changed. <br>
+     * It is called from {@link GrindstoneMenu#createResult()}. <br>
+     * If the event is canceled, vanilla behavior will not run, and the output will be set to {@link ItemStack#EMPTY}. <br>
```

Comment:
Look at other events for the style here, it should say something like "This event is {@link Cancelable}"